upbc/protoc-gen-upb.cc

```diff
@@ -369,6 +329,346 @@ std::string CTypeConst(const protobuf::FieldDescriptor* field) {
+  uintptr_t val = reinterpret_cast<uintptr_t>(data);
+  assert((val & kMask) == 0);
+  upb_MiniTable_Sub sub;
+  sub.submsg = reinterpret_cast<upb_MiniTable*>(val | tag);
+  return sub;
+}
+
+bool FilePlatformLayout::IsNull(upb_MiniTable_Sub sub) {
+  return reinterpret_cast<uintptr_t>(sub.subenum) == 0;
+}
+
+std::string FilePlatformLayout::GetSub(upb_MiniTable_Sub sub) {
+  uintptr_t as_int = reinterpret_cast<uintptr_t>(sub.submsg);
+  const char* str = reinterpret_cast<const char*>(as_int & ~SubTag::kMask);
+  switch (as_int & SubTag::kMask) {
+    case SubTag::kMessage:
+      return absl::Substitute("{.submsg = &$0}", str);
+    case SubTag::kEnum:
+      return absl::Substitute("{.subenum = &$0}", str);
+    default:
+      return std::string("{.submsg = NULL}");
+  }
+  return std::string("ERROR in WriteSub");
+}
+
+void FilePlatformLayout::SetSubTableStrings() {
+  for (const auto& pair : table_map_) {
+    upb_MiniTable* mt = pair.second;
+    for (const auto& f : FieldNumberOrder(pair.first)) {
+      upb_MiniTable_Field* mt_f = const_cast<upb_MiniTable_Field*>(
+          upb_MiniTable_FindFieldByNumber(mt, f->number()));
+      assert(mt_f);
+      upb_MiniTable_Sub sub = PackSubForField(f, mt_f);
+      if (IsNull(sub)) continue;
+      *const_cast<upb_MiniTable_Sub*>(&mt->subs[mt_f->submsg_index]) = sub;
+    }
+  }
+}
+
+upb_MiniTable_Sub FilePlatformLayout::PackSubForField(
+    const protobuf::FieldDescriptor* f, const upb_MiniTable_Field* mt_f) {
+  if (mt_f->submsg_index == kUpb_NoSub) {
+    return PackSub(nullptr, SubTag::kNull);
+  } else if (f->message_type()) {
+    return PackSub(AllocStr(MessageInit(f->message_type())), SubTag::kMessage);
+  } else {
+    ABSL_ASSERT(f->enum_type());
+    return PackSub(AllocStr(EnumInit(f->enum_type())), SubTag::kEnum);
+  }
+}
+
+const char* FilePlatformLayout::AllocStr(const std::string& str) {
+  char* ret =
+      static_cast<char*>(upb_Arena_Malloc(arena_.ptr(), str.size() + 1));
+  memcpy(ret, str.data(), str.size());
+  ret[str.size()] = '\0';
+  return ret;
+}
+
+void FilePlatformLayout::BuildMiniTables(const protobuf::FileDescriptor* fd) {
+  for (const auto& m : SortedMessages(fd)) {
+    table_map_[m] = MakeMiniTable(m);
+  }
+  ResolveIntraFileReferences();
+  SetSubTableStrings();
+}
+
+void FilePlatformLayout::BuildExtensions(const protobuf::FileDescriptor* fd) {
+  auto sorted = SortedExtensions(fd);
+  upb::Status status;
+  for (const auto& f : sorted) {
+    upb::MtDataEncoder e;
+    e.StartMessage(0);
+    e.PutField(static_cast<upb_FieldType>(f->type()), f->number(),
+               GetFieldModifiers(f));
+    upb_MiniTable_Extension& ext = extension_map_[f];
+    upb_MiniTable_Sub sub;
+    bool ok = upb_MiniTable_BuildExtension(e.data().data(), e.data().size(),
+                                           &ext, sub, status.ptr());
+    if (!ok) {
+      fprintf(stderr, "Error building mini-table: %s\n",
```

Comment:
not sure how much this buys you given the assert.  Also, the fact that it blindly continues is a bit troubling

Reply:
Bazel makes it very hard to use an asserts-enabled build of a code generator during the build. I'm sort of biding my time until we have ABSL logging and can use `CHECK()`.

Reply:
can you add a TODO?